optimizer/lib/transformers/BlurImagePlaceholders.js

```diff
@@ -0,0 +1,158 @@
+    return Promise.all(promises);
+  }
+
+  /**
+   * Adds a child image that is a blurry placeholder.
+   * @param tree A parse5 treeAdapter.
+   * @param {String} src The image that the bitmap is based on.
+   * @return {!Promise} A promise that signifies that the img has been updated
+   * to have correct attributes to be a blurred placeholder.
+   * @private
+   */
+  addBlurryPlaceholder_(tree, src) {
+    const img = tree.createElement('img');
+    img.attribs.src = src;
+    img.attribs.class = 'i-amphtml-blur';
+    img.attribs.placeholder = '';
+    return this.getDataURI_(img).then(() => {
+      return img;
+    }).catch(err => {
+      console.log(err);
+    });
+  }
+
+  /**
+   * Creates the bitmap in a dataURI format.
+   * @param {Node} img The DOM element that needs a dataURI for the
+   * placeholder.
+   * @return {!Promise} A promise that is resolved once the img's src is updated
+   * to be a dataURI of a bitmap.
+   * @private
+   */
+  getDataURI_(img) {
+    const bitMapDims = this.getBitmapDimensions_(img);
+    return this.createBitmap_(img, bitMapDims.width, bitMapDims.height).then(dataURI => {
+      img.attribs.src = dataURI;
+    });
+  }
+
+  /**
+   * Calculates the correct dimensions for the bitmap.
+   * @param {Node} img The DOM element that will need a bitmap.
+   * placeholder.
+   * @return {Record} The aspect ratio of the bitmap of the image.
+   * @private
+   */
+  getBitmapDimensions_(img) {
+    // Gets the original aspect ratio of the image.
+    const imgDims = sizeOf(img.attribs.src);
+    const imgWidth = imgDims.width;
+    const imgHeight = imgDims.height;
+    // Aims for a bitmap of ~60 pixels (w * h = ~60).
+    const bitmapPixelAmt = 60;
+    // Gets the ratio of the width to the height. (r = w0 / h0 = w / h)
+    const ratioWH = imgWidth / imgHeight;
+    // Express the width in terms of height by multiply the ratio by the
+    // height. (h * r = (w / h) * h)
+    // Plug this representation of the width into the original equation.
+    // (h * r * h = ~60).
+    // Divide the bitmap size by the ratio to get the all expressions using
+    // height on one side. (h * h = ~60 / r)
+    let bitmapHeight = bitmapPixelAmt / ratioWH;
+    // Take the square root of the height instances to find the singular value
+    // for the height. (h = sqrt(~60 / r))
+    bitmapHeight = Math.sqrt(bitmapHeight);
+    // Divide the goal total pixel amount by the height to get the width.
+    // (w = ~60 / h).
+    const bitmapWidth = bitmapPixelAmt / bitmapHeight;
+    return {width: Math.round(bitmapWidth), height: Math.round(bitmapHeight)};
+  }
+
+  /**
+   * Reads and transforms an image to a blurry placeholder format.
+   * @param {Node} img The DOM element that will need a bitmap.
+   * placeholder.
+   * @return {!Promise} A promise that is resolved when the image has been read
+   * and transformed.
+   * @private
+   */
+  createBitmap_(img, width, height) {
+    return jimp.read(img.attribs.src)
```

Comment:
Am I right that this will only work for absolute image URLs (and the images need to be deployed already)? 

If so, we should find a way to make this configurable, either by:

* passing a image host via parameter
* passing a function that will resolve the image (path) from a given `src` value

Reply:
So currently, it does only support absolute image URLs, but we will be exploring ways to account for relative file paths. Through a cursory search, I wasn't able to find another place in the code base to reference for this, so if there is, then please let me know :)

Reply:
You could pass either a string as transformer params:

```
const params = {
  imagePrefix: 'https://example.com' or '../public/'
}
optimizer.transformHtml(html, params)
```

or a function that returns the read input from the src value:

```
const params = {
  imagePrefix: src => 'https://example.com/' + src
}
```

and access it in the transformer via:

```
transform(tree, params) {
  resolveFile(src, params.imagePrefix)
  ...
}
```

Reply:
Ahh I see! Thank you!